File: oop/lab1/cpp/app/main.cpp
#include <iostream>
#include <vector>

#include <stringConverter.hpp>
#include <primeNumbers.hpp>

/**
 * @brief Main function
 *
 * @param argc - number of arguments
 * @param argv - array of arguments
 * @return int
 */
int main(const int argc, const char *const argv[])
{
    if (argc < 2)
    {
        std::cerr << "Usage: " << argv[0] << " <arg1> <arg2> ..." << std::endl;
        return 1;
    }

    PrimeNumbers* p;

    try
    {
        const auto initArgument = StringConverter::stringToVal<int>(std::string(argv[1]));
        p = new PrimeNumbers(initArgument);
    }
    catch (const std::invalid_argument &e)
    {
        std::cerr << "Invalid Argument Error: " << e.what() << " should be greater than 1" << std::endl;
        return 1;
    }

    std::cout << "Successfully initialized PrimeNumbers object of count " << std::to_string(p->getCount()) << std::endl;

    for (std::size_t i = 2; i < static_cast<std::size_t>(argc); i++)
    {
        try
        {
            const auto argNumber = StringConverter::stringToVal<int>(std::string(argv[i]));
            const auto prime = p->getPrime(argNumber);

            std::cout << std::to_string(argNumber) << " --> " << std::to_string(prime) << std::endl;
        }
        catch (const std::invalid_argument &e)
        {
            std::cerr << "Invalid Argument Error: " << e.what() << std::endl;
        }
        catch (const std::out_of_range &e)
        {
            std::cerr << "Out of Range Error: " << e.what() << std::endl;
        }
    }

    delete p;

    return 0;
}

File: oop/lab1/java/Main.java
public final class Main {
    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java Main <arg1> <arg2> ...");
            System.exit(1);
        }

        PrimeNumbers p = null;

        try {
            int initArgument = Integer.parseInt(args[0]);
            p = new PrimeNumbers(initArgument);
        } catch (IllegalArgumentException e) {
            System.err.println("Invalid Argument Error: " + e.getMessage());
            System.exit(1);
        }

        System.out.println("Successfully initialized PrimeNumbers object of count " + p.getCount());

        for (int i = 1; i < args.length; i++) {
            try {
                int argNumber = Integer.parseInt(args[i]);
                int prime = p.getPrime(argNumber);

                System.out.println(argNumber + " --> " + prime);
            } catch (NumberFormatException e) {
                System.err.println("Invalid Argument Error: " + e.getMessage());
            } catch (IndexOutOfBoundsException e) {
                System.err.println("Out of Range Error: " + e.getMessage());
            } catch (IllegalArgumentException e) {
                System.err.println("Invalid Argument Error: " + e.getMessage());
            }
        }
    }
}